src/runtime/rposix.r

```diff
@@ -1576,6 +1582,297 @@ int fd;
+	// TLS1_2_VERSION, TLS1_3_VERSION for TLS and DTLS1_VERSION, DTLS1_2_VERSION for DTLS.
+#if !defined(MacOS) && (OPENSSL_VERSION_NUMBER < 0x10100000L)
+#define TLS1_VERSION 10
+#define TLS1_1_VERSION 11
+#define TLS1_2_VERSION 12
+#define TLS1_3_VERSION 13
+
+#define DTLS1_VERSION 20
+#define DTLS1_2_VERSION 22
+
+#endif
+	int ver;
+        if (strcmp(proto, "TLS1.3") == 0)
+          ver = TLS1_3_VERSION;
+	else if (strcmp(proto, "TLS1.2") == 0)
+	  ver = TLS1_2_VERSION;
+	else if (strcmp(proto, "TLS1.1") == 0)
+	  ver = TLS1_1_VERSION;
+	else if (strcmp(proto, "TLS1.0") == 0)
+	  ver = TLS1_VERSION;
+	/*	else if (strcmp(proto, "SSL3.0") == 0)
+		ver = SSL3_VERSION; */
+	else if (strcmp(proto, "DTLS1.2") == 0)
+	  ver = DTLS1_2_VERSION;
+	else if (strcmp(proto, "DTLS1.0") == 0)
+	  ver = DTLS1_VERSION;
+        else {
+	  set_ssl_context_errortext(1308, proto);
+	  return NULL;
+	}
+
+
+	if (count == 2) {
+#if !defined(MacOS) && (OPENSSL_VERSION_NUMBER < 0x10100000L)
+	  switch (ver) {
+	  case TLS1_2_VERSION: old_ssl_flags |= SSL_OP_NO_TLSv1 | SSL_OP_NO_TLSv1_1 |
+	      SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3; break;
+	  case TLS1_1_VERSION: old_ssl_flags |= SSL_OP_NO_TLSv1 |
+	      SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3; break;
+	  case TLS1_VERSION: old_ssl_flags |= SSL_OP_NO_SSLv2 | SSL_OP_NO_SSLv3; break;
+	  default:
+	    set_ssl_context_errortext(1308, proto);
+	    return NULL;
+	  }
+	  SSL_CTX_set_options(ctx, old_ssl_flags);
+#else
+	  if (SSL_CTX_set_min_proto_version(ctx, ver) != 1) {
+	    set_ssl_context_errortext(1301, proto);
+	    return NULL;
+	  }
+#endif
+	}
+	else {
+#if !defined(MacOS) && (OPENSSL_VERSION_NUMBER < 0x10100000L)
+	  switch (ver) {
+	  case TLS1_2_VERSION: break;
+	  case TLS1_1_VERSION: old_ssl_flags |= SSL_OP_NO_TLSv1_2; break;
+	  case TLS1_VERSION: old_ssl_flags |= SSL_OP_NO_TLSv1_1 | SSL_OP_NO_TLSv1_2; break;
+	  default:
+	    set_ssl_context_errortext(1308, proto);
+	    return NULL;
+	    }
+	  SSL_CTX_set_options(ctx, old_ssl_flags);
+#else
+	  if (SSL_CTX_set_max_proto_version(ctx, ver) != 1) {
+	    set_ssl_context_errortext(1301, proto);
+	    return NULL;
+	  }
+#endif
+	}
+      }
+   } while (--count>0);
+
+   if (ca_file != NULL || ca_dir != NULL) {
+     if (SSL_CTX_load_verify_locations(ctx, ca_file, ca_dir) != 1) {
+       set_ssl_context_errortext(1305, ca_file);
+       return NULL;
+     }
+   }
+   if (verifyPeer == NULL) {
+     if ((type == TLS_CLIENT) || (type == DTLS_CLIENT)) {
+       // cannot fail
+       SSL_CTX_set_verify(ctx, SSL_VERIFY_PEER,  NULL);
+     }
+   } else {
+     if (strcmp(verifyPeer, "yes") == 0) {
+       if ((type == TLS_CLIENT) || (type == DTLS_CLIENT))
+	 SSL_CTX_set_verify(ctx, SSL_VERIFY_PEER,  NULL);
+       else
+	 SSL_CTX_set_verify(ctx,  SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT,  NULL);
+     } else if (strcmp(verifyPeer, "no") != 0) {
+       set_errortext_with_val(1302, verifyPeer);
```

Comment:
Add 1302 to Appendix A of the Unicon Book.  Where are errors 1300 and 1301?

Reply:
done